service/history/archival_queue_task_executor.go

```diff
@@ -0,0 +1,343 @@
+	if err != nil {
+		return err
+	}
+	defer func() {
+		mutableState.Release(err)
+	}()
+
+	taskGenerator := workflow.NewTaskGenerator(
+		e.shardContext.GetNamespaceRegistry(),
+		mutableState,
+		e.shardContext.GetConfig(),
+	)
+	closeTime, err := mutableState.GetWorkflowCloseTime(ctx)
+	if err != nil {
+		return err
+	}
+	err = taskGenerator.GenerateDeleteHistoryEventTask(*closeTime, true)
+	if err != nil {
+		return nil
+	}
+	err = e.shardContext.AddTasks(ctx, &persistence.AddHistoryTasksRequest{
+		ShardID:     e.shardContext.GetShardID(),
+		NamespaceID: task.GetNamespaceID(),
+		WorkflowID:  task.WorkflowID,
+		RunID:       task.RunID,
+		Tasks:       mutableState.PopTasks(),
+	})
+	return err
+}
+
+// lockedMutableState is a wrapper around mutable state that includes the last write version of the mutable state and
+// a function to release this mutable state's context when we're done with it.
+type lockedMutableState struct {
+	// MutableState is the mutable state that is being wrapped. You may call any method on this object safely since
+	// the state is locked.
+	workflow.MutableState
+	// LastWriteVersion is the last write version of the mutable state. We store this here so that we don't have to
+	// call GetLastWriteVersion() on the mutable state object again.
+	LastWriteVersion int64
+	// Release is a function that releases the context of the mutable state. This function should be called when
+	// you are done with the mutable state.
+	Release cache.ReleaseCacheFunc
+}
+
+// newLockedMutableState returns a new lockedMutableState with the given mutable state,
+// last write version and release function
+func newLockedMutableState(
+	mutableState workflow.MutableState,
+	version int64,
+	releaseFunc cache.ReleaseCacheFunc,
+) *lockedMutableState {
+	return &lockedMutableState{
+		MutableState:     mutableState,
+		LastWriteVersion: version,
+		Release:          releaseFunc,
+	}
+}
+
+// loadAndVersionCheckMutableState loads the mutable state for the given task and performs a version check to verify
+// that the mutable state's last write version is equal to the task's version. If the version check fails then
+// the mutable state is released and an error is returned. This can happen if this task was already processed.
+func (e *archivalQueueTaskExecutor) loadAndVersionCheckMutableState(
+	ctx context.Context,
+	task tasks.Task,
+) (lockedMutableState *lockedMutableState, err error) {
+	weContext, release, err := getWorkflowExecutionContextForTask(ctx, e.workflowCache, task)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		// If we return an error, the caller will not release the mutable state, so we need to do it here.
+		if err != nil {
+			release(err)
+		}
+		// If we don't return an error, the caller will release the mutable state, so we don't need to do it here.
+	}()
+
+	mutableState, err := weContext.LoadMutableState(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if mutableState == nil {
```

Comment:
Could this happen if a workflow is deleted before the archival task?

Reply:
Changed to drop the task and log a warning. Also added a test case for this branch.

Reply:
return nil error to drop the task?

Reply:
I return a const error here and then I convert it to nil further up the stack. If I return nil here, then execution will continue because this is a helper method.